sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/crypto/CryptoAesAndEtmHandler.java

```diff
@@ -0,0 +1,165 @@
+package com.microsoft.appcenter.utils.crypto;
+
+import android.content.Context;
+import android.os.Build;
+import android.security.keystore.KeyGenParameterSpec;
+import android.security.keystore.KeyProperties;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.RequiresApi;
+import androidx.annotation.VisibleForTesting;
+
+import java.nio.ByteBuffer;
+import java.security.InvalidKeyException;
+import java.security.KeyStore;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Calendar;
+
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.IvParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+
+import static com.microsoft.appcenter.utils.crypto.CryptoConstants.AES_KEY_SIZE;
+import static com.microsoft.appcenter.utils.crypto.CryptoConstants.ANDROID_KEY_STORE;
+import static com.microsoft.appcenter.utils.crypto.CryptoConstants.ENCRYPT_KEY_LIFETIME_IN_YEARS;
+import static javax.crypto.Cipher.DECRYPT_MODE;
+import static javax.crypto.Cipher.ENCRYPT_MODE;
+
+public class CryptoAesAndEtmHandler implements CryptoHandler {
+
+    private static final int ENCRYPTION_KEY_LENGTH = 32;
+    private static final int AUTHENTICATION_KEY_LENGTH = 16;
+
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    @Override
+    public String getAlgorithm() {
+        return CryptoConstants.CIPHER_AES + "/" + AES_KEY_SIZE + "/" + KeyProperties.KEY_ALGORITHM_HMAC_SHA256;
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    @Override
+    public void generateKey(CryptoUtils.ICryptoFactory cryptoFactory, String alias, Context context) throws Exception {
+        Calendar writeExpiry = Calendar.getInstance();
+        writeExpiry.add(Calendar.YEAR, ENCRYPT_KEY_LIFETIME_IN_YEARS);
+        CryptoUtils.IKeyGenerator keyGenerator = cryptoFactory.getKeyGenerator(KeyProperties.KEY_ALGORITHM_HMAC_SHA256, ANDROID_KEY_STORE);
+        keyGenerator.init(new KeyGenParameterSpec.Builder(alias, KeyProperties.PURPOSE_SIGN)
+                        .setKeyValidityForOriginationEnd(writeExpiry.getTime())
+                        .build());
+        keyGenerator.generateKey();
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    @Override
+    public byte[] encrypt(CryptoUtils.ICryptoFactory cryptoFactory, int apiLevel, KeyStore.Entry keyStoreEntry, byte[] input) throws Exception {
+
+        // Get secure key and subkeys.
+        byte[] encryptionSubkey = getSubkey(((KeyStore.SecretKeyEntry) keyStoreEntry).getSecretKey(), ENCRYPTION_KEY_LENGTH);
```

Comment:
How heavy is `((KeyStore.SecretKeyEntry) keyStoreEntry).getSecretKey()` operation? Shouldn't we store the value to var and use it in both cases?